martin/src/geojson/mod.rs

```diff
@@ -0,0 +1,117 @@
+use std::fmt::{Debug, Formatter};
+use std::io;
+use std::path::PathBuf;
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use geojson_vt_rs::{GeoJSONVT, Options};
+use martin_tile_utils::{Format, TileCoord, TileInfo};
+use prost::Message as _;
+use std::fs::File;
+use tilejson::TileJSON;
+use tilejson::tilejson;
+use tokio::sync::Mutex;
+
+use crate::file_config::FileError::{self, InvalidMetadata, IoError};
+use crate::file_config::FileResult;
+use crate::geojson::mvt::LayerBuilder;
+use crate::source::{TileData, TileInfoSource, UrlQuery};
+use crate::{MartinResult, Source};
+
+mod config;
+mod mvt;
+
+pub use config::GeoJsonConfig;
+
+pub struct GeoJsonSource {
+    id: String,
+    path: PathBuf,
+    inner: Arc<Mutex<GeoJSONVT>>,
+    tilejson: TileJSON,
+    tile_info: TileInfo,
+}
+
+impl Debug for GeoJsonSource {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "GeoJsonSource {{ id: {}, path: {:?} }}",
+            self.id, self.path,
+        )
+    }
+}
+
+impl GeoJsonSource {
+    fn new(id: String, path: PathBuf) -> FileResult<Self> {
+        let tile_info = TileInfo::new(Format::Mvt, martin_tile_utils::Encoding::Uncompressed);
+        let geojson_file =
+            File::open(&path).map_err(|e: std::io::Error| FileError::IoError(e, path.clone()))?;
+
+        // TODO: better error handling
+        let geojson = geojson::GeoJson::from_reader(geojson_file)
+            .map_err(|e| FileError::InvalidFilePath(path.clone()))?;
+
+        // TODO: see if options default is goood enough
+        let inner = GeoJSONVT::from_geojson(&geojson, &Options::default());
+
+        let tilejson = tilejson! {
+            tiles: vec![],
+            minzoom: 0,
+            maxzoom: 18,
+        };
+
+        return Ok(Self {
+            id,
+            path,
+            inner: Arc::new(Mutex::new(inner)),
+            tilejson,
+            tile_info,
+        });
+    }
+}
+
+#[async_trait]
+impl Source for GeoJsonSource {
+    fn get_id(&self) -> &str {
+        &self.id
+    }
+
+    fn get_tilejson(&self) -> &TileJSON {
+        &self.tilejson
+    }
+
+    fn get_tile_info(&self) -> TileInfo {
+        self.tile_info
+    }
+
+    fn clone_source(&self) -> TileInfoSource {
+        // TODO: implement clone for GeoJsonSource (GeoJSONVT is not cloneable, which is a blocker)
+        Box::new(Self::new(self.id.clone(), self.path.clone()).unwrap())
+    }
+
+    fn benefits_from_concurrent_scraping(&self) -> bool {
+        // TODO: figure out if this is a good idea
+        false
```

Comment:
If we switch to `RWLock`, the answer should be `true`. If we cannot, the answer is `false` with a reference to the `Mutex`

```suggestion
        // Extracting information from JSON is fairly CPU intensitve
        // => we can use more IO by using more threads
        true
```

Reply:
The problem is `GeoJSONVT`'s `get_tile` method takes a `&mut self`, so I don't think it would make a huge difference. The `GeoJSONVT` struct internally caches tiles in a HashMap, as tiles at higher zoom are created by splitting tiles at lower zooms. The `&mut self` makes it so that we can get one tile at a time, as far as I understand.

I can use [`geojson_to_tile()`](https://docs.rs/geojson-vt-rs/0.1.1/geojson_vt_rs/fn.geojson_to_tile.html) directly but that means losing the caching. Another option would be to move the RwLock into `GeoJSONVT` itself -- not sure 100%.

@maxammann is my understanding correct? what do you think?

Reply:
Yea, using `GeoJSONVT` then does not quite work for us then.

We already have a multithreaded cache outside of this function
=> lets use [`geojson_to_tile`](https://docs.rs/geojson-vt-rs/0.1.1/geojson_vt_rs/fn.geojson_to_tile.html) with either an `Arc` or `Arc<RWLock<GeoJSON>>`
=> nicely paralell.

Max does not need paralellisability as wasm does not have good multithreading support.